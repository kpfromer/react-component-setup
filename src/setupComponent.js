import React from "react";
import { mount, shallow } from "enzyme";

// todo: extract into a simple npm package
const setupComponent = (Component, elementsToFind = [], defaultProps = {}) => {
  const assembleProps = props => ({...defaultProps, ...props});

  const constructComponent = (props, construct) => {

    const wrapper = construct(<Component {...assembleProps(props)}/>);

    const foundElements = {};

    elementsToFind.forEach(element => {
      foundElements[element.name] = wrapper.find(element.query)
    });

    return {
      wrapper,
      ...foundElements
    }
  };

  return {
    mount: (props = {}) => constructComponent(props, mount),
    shallow: (props = {}) => constructComponent(props, shallow)
  }
};

export default setupComponent;