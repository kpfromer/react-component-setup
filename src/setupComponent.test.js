import React, { Component } from 'react';
import setupComponent from "./setupComponent";

class TestComponent extends Component {
  render() {
    return <h1>{this.props.coolTitle}</h1>;
  }
}

class ChildComponent extends Component {
  render() {
    return <p>{this.props.data}</p>;
  }
}

class ParentComponent extends Component {
  render() {
    return <div>
      <h1>{this.props.title}</h1>
      <ChildComponent data={this.props.data} />
    </div>
  }
}

class MultiItems extends Component {
  render() {
    return (
      <div>
        <h1>h1</h1>
        <h2>hello world</h2>
        <div className="coolClass">
          cool is what I am
        </div>
      </div>
    );
  }
}

describe('setupComponent', () => {
  describe('shallow', () => {
    let shallow;

    beforeEach(() => {
      const { shallow: testShallow } = setupComponent(TestComponent, [], {
        coolTitle: 'i am a default prop'
      });

      shallow = testShallow;
    });

    it('constructs a shallow component', () => {
      const { wrapper } = shallow();
      expect(wrapper).toMatchSnapshot();
    });

    it('overrides default props with function props', () => {
      const { wrapper } = shallow({
        coolTitle: 'I am a cool overridden prop!!'
      });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('mount', () => {
    let mount;

    beforeEach(() => {
      const { mount: testMount } = setupComponent(ParentComponent, [], {
        title: 'default title',
        data: 'default data'
      });

      mount = testMount;
    });

    it('constructs a mount component', () => {
      const { wrapper } = mount();
      expect(wrapper).toMatchSnapshot();
    });

    it('overrides default props with function props', () => {
      const { wrapper } = mount({
        data: 'overridden data'
      });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('given `elementsToFind`', () => {
    let shallow;

    beforeEach(() => {
      const { shallow: testShallow } = setupComponent(MultiItems, [
        {
          name: 'coolClass',
          query: '.coolClass'
        }
      ]);

      shallow = testShallow;
    });

    it('returns the element', () => {
      expect(shallow().coolClass).toMatchSnapshot();
    });
  });
});