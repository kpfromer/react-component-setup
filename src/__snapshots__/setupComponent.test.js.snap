// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupComponent given \`elementsToFind\` returns the element 1`] = `
<div
  className="coolClass"
>
  cool is what I am
</div>
`;

exports[`setupComponent mount constructs a mount component 1`] = `
<div>
  <h1>
    default title
  </h1>
  <p>
    default data
  </p>
</div>
`;

exports[`setupComponent mount overrides default props with function props 1`] = `
<div>
  <h1>
    default title
  </h1>
  <p>
    overridden data
  </p>
</div>
`;

exports[`setupComponent shallow constructs a shallow component 1`] = `
<h1>
  i am a default prop
</h1>
`;

exports[`setupComponent shallow overrides default props with function props 1`] = `
<h1>
  I am a cool overridden prop!!
</h1>
`;
